File: 8-string_to_integer.js
/**
 * Converts a string to a 32-bit signed integer following the atoi rules.
 * @param {string} s - The input string.
 * @return {number} - The converted integer.
 */
function myAtoi(s) {
    // Step 1: Trim leading and trailing spaces
    s = s.trim();
    
    if (s.length === 0) return 0; // Return 0 if string is empty after trimming
    
    let index = 0, sign = 1, result = 0;
    
    // Step 2: Check for sign
    if (s[index] === '+' || s[index] === '-') {
        sign = s[index] === '-' ? -1 : 1;
        index++; // Move to the next character
    }
    
    // Step 3: Convert numeric characters to an integer
    while (index < s.length && s[index] >= '0' && s[index] <= '9') {
        result = result * 10 + (s[index] - '0');
        index++;
    }
    
    // Step 4: Apply sign
    result *= sign;
    
    // Step 5: Clamp the value to 32-bit signed integer range
    const INT_MAX = 2 ** 31 - 1;
    const INT_MIN = -(2 ** 31);
    
    if (result > INT_MAX) return INT_MAX;
    if (result < INT_MIN) return INT_MIN;
    
    return result;
}

// Example test cases
console.log(myAtoi("42")); // Output: 42
console.log(myAtoi("   -42")); // Output: -42
console.log(myAtoi("4193 with words")); // Output: 4193
console.log(myAtoi("words and 987")); // Output: 0
console.log(myAtoi("-91283472332")); // Output: -2147483648 (clamped to INT_MIN)

File: 8-string_to_integer.py
def myAtoi(s: str) -> int:
    # Step 1: Strip leading whitespaces from the input string
    s = s.lstrip()
    
    # Step 2: Handle edge case where the string becomes empty after trimming
    if not s:
        return 0
    
    # Step 3: Determine the sign of the number (+ or -)
    sign = 1
    if s[0] == '-':
        sign = -1
        s = s[1:]  # Remove the sign character
    elif s[0] == '+':
        s = s[1:]  # Remove the sign character
    
    # Step 4: Initialize result variable to store the number
    result = 0
    
    # Step 5: Process each character and convert digits to an integer
    for char in s:
        # If the character is a digit, add it to the result
        if char.isdigit():
            result = result * 10 + int(char)
        else:
            break  # Stop when we encounter a non-digit character
    
    # Step 6: Apply the sign to the result
    result *= sign
    
    # Step 7: Handle overflow cases, the result must be in the range [−2^31, 2^31 − 1]
    # Python's int type can handle large integers, but we need to clamp it to the 32-bit range.
    int_min, int_max = -2**31, 2**31 - 1
    if result < int_min:
        return int_min
    elif result > int_max:
        return int_max
    
    return result
